dali/pipeline/data/tensor_list.h

```diff
@@ -322,18 +336,66 @@ class DLL_PUBLIC TensorList : public Buffer<Backend> {
   }
 
   /**
-   * @brief Returns a Tensor which shares the data
-   * with this TensorList. The tensor obtained
-   * through this function stays valid for the lifetime
-   * of the parent TensorList.
+   * @brief Returns a Tensor view with given shape or nullptr if no
+   * such exists
+   */
+  inline Tensor<Backend> * GetViewWithShape(const vector<Index> &shape) {
+    for (auto &t : tensor_views_) {
+      if (t->shape() == shape) {
+        return t.get();
+      }
+    }
+    return nullptr;
+  }
+
+  /**
+   * @brief Returns a pointer to Tensor which shares the data
+   * with this TensorList and give it the provided shape.
+   * Tensor list owns the memory. The tensor obtained through
+   * this function stays valid for as long as TensorList data is unchanged.
    */
-  Tensor<Backend> * AsTensor() {
-    if (tensor_view_ == nullptr) {
-      tensor_view_ = new Tensor<Backend>();
-      tensor_view_->ShareData(this);
+  DLL_PUBLIC inline Tensor<Backend> * AsReshapedTensor(const vector<Index> &new_shape) {
+    auto t = GetViewWithShape(new_shape);
+    if (t) {
+      return t;
     }
 
-    return tensor_view_;
+    // need to create a new view
+    auto new_view = std::unique_ptr< Tensor<Backend> >(new Tensor<Backend>());
```

Comment:
List of `unique_ptr` has little advatage over list of values if we don't allow transfer of ownership outside the list.

Reply:
Done